### PR TITLE
Stability fix

### PR DESCRIPTION
**Pull request by MarkMcWire**
Fixes the memory leak from Version 1.3, so the game don't crash. But the mod still needs 20 GB of additional RAM on Linux.

### CODE REVIEW

**Comment by Gregory365** [comment]
Over 3000 lines changed!?

**Comment by Gregory365** [comment]
Can you split the whitespace refactor and the memory leak fix, at least into separate commits so I can review what has been changed?
Thank you for looking into this issue, it's been bugging me for months.

**Comment by MarkMcWire** [comment]
> Can you split the whitespace refactor and the memory leak fix, at least into separate commits so I can review what has been changed? Thank you for looking into this issue, it's been bugging me for months.

How? My editor automatically converts spaces to tabs.

**Comment by Gregory365** [comment]
Ok nvm, TIL a way view the changes ignoring the whitespace by appending '?w=1' to the end of the URL
https://github.com/Gregory365/TPF2-Timetables/pull/75/files?w=1

File: res/config/game_script/timetable_gui.lua
Comment (Gregory365) [inline res/config/game_script/timetable_gui.lua:93]: Why is `timetableChanged` being set to `false`?

File: res/config/game_script/timetable_gui.lua
Comment (Gregory365) [inline res/config/game_script/timetable_gui.lua:1156]: Question for my understanding:
Can `guiUpdate` be in a separate file to the `update`, `load`, and 'save' functions?
Will all those functions still get called in the same way?

File: res/scripts/celmi/timetables/timetable_helper.lua
Comment (Gregory365) [inline res/scripts/celmi/timetables/timetable_helper.lua:451]: From reading the transportfever.net forum I now understand that calls to the API should be done on separate lines. Does this also extend to the `api.type.ComponentType.[type_name]`?

Also I am not sure if the `.gameTime` should be split to a separate line as well. See https://github.com/IncredibleHannes/TPF2-Timetables/issues/50
```
local timeComponent = api.engine.getComponent(apiWorld, apiGameTime)
local time = timeComponent.gameTime
```

File: res/scripts/celmi/timetables/timetable.lua
Comment (Gregory365) [inline res/scripts/celmi/timetables/timetable.lua:62]: For my understanding, can you clarify why setting this to nil is necessary?

File: res/config/game_script/timetable_engine.lua
Comment (Gregory365) [inline res/config/game_script/timetable_engine.lua:14]: Are `UIStrings` needed in both timetable_gui.lua and timetable_engine.lua?

**Review by Gregory365 — CHANGES_REQUESTED** [review]
Thanks for making these changes, they look good. I've got some questions, most of them just so I understand the reasoning behind the changes.

File: res/config/game_script/timetable_gui.lua
Comment (MarkMcWire) [reply res/config/game_script/timetable_gui.lua:93]: Because this function only fills the station list in the GUI without changing the content of the timetable. This way I prevent the timetable object from being created again when the station list is displayed.

File: res/config/game_script/timetable_gui.lua
Comment (MarkMcWire) [reply res/config/game_script/timetable_gui.lua:1156]: Yes. We discussed this in the TPF forum. It is better for the stability of the script if GUI elements and other function calls are created separately. The game itself creates two separate threads for this in the background anyway.

File: res/config/game_script/timetable_engine.lua
Comment (MarkMcWire) [reply res/config/game_script/timetable_engine.lua:14]: This is experimental. I wanted to remove any remnants of previous instances from the object. This way it is basically created from scratch.

File: res/scripts/celmi/timetables/timetable.lua
Comment (MarkMcWire) [reply res/scripts/celmi/timetables/timetable.lua:62]: This is experimental. I wanted to remove any remnants of previous instances from the object. This way it is basically created from scratch.

File: res/config/game_script/timetable_gui.lua
Comment (Gregory365) [reply res/config/game_script/timetable_gui.lua:93]: Ok I agree. Should this line be omitted rather than set to false? Something else might change at the same time that does cause the timetable to change.

File: res/config/game_script/timetable_engine.lua
Comment (MarkMcWire) [reply res/config/game_script/timetable_engine.lua:14]: I must have overlooked something there.

File: res/config/game_script/timetable_gui.lua
Comment (MarkMcWire) [reply res/config/game_script/timetable_gui.lua:93]: > Ok I agree. Should this line be omitted rather than set to false? Something else might change at the same time that does cause the timetable to change.

To remove it would maybe be better. Your right.

File: res/scripts/celmi/timetables/timetable_helper.lua
Comment (MarkMcWire) [reply res/scripts/celmi/timetables/timetable_helper.lua:451]: I would not change it, the ways its now. This is only a call of this object not call up of the api itself.

**Review by Gregory365 — APPROVED** [review]

**Comment by Gregory365** [comment]
I've put a request in the timetables discord server for people to test this branch. I'll wait for feedback from that test before merging this in and pushing the update to steam workshop.

**Pull request merged**
